src/logging.h

```diff
@@ -163,4 +166,89 @@ static inline void LogPrint(const BCLog::LogFlags& category, const Args&... args
     }
 }
 
+
+namespace BCLog {
+
+//! RAII-style object that outputs timing information to logs.
+template <typename TimeType>
+class Timer
+{
+public:
+    //! Log prefix; usually the name of the function this was created in.
+    std::string m_prefix{};
+
+    //! A descriptive message of what is being timed.
+    std::string m_title{};
+
+    //! Forwarded on to LogPrint if specified - has the effect of only
+    //! outputing the timing log when a particular debug= category is specified.
+    BCLog::LogFlags m_log_category{};
+
+    //! If log_category is left as the default, end_msg will log unconditionally
+    //! (instead of being filtered by category).
+    Timer(
+        std::string prefix,
+        std::string end_msg,
+        BCLog::LogFlags log_category = BCLog::LogFlags::ALL) :
+            m_prefix(std::move(prefix)),
+            m_title(std::move(end_msg)),
+            m_start_t(GetTime<std::chrono::microseconds>()),
+            m_log_category(log_category)
+    {
+        this->Log(strprintf("%s started", m_title), false);
```

Comment:
Maybe wait to initialize m_start_t until after the Log call, so the elapsed time doesn't include the time spent logging. This would also let you eliminate the `show_elapsed_time` bool arguments (could just check `m_start_t != 0` instead).

Reply:
Oh that's a good point. Will do.